Have createwalletdescriptor auto-detect an unused(KEY)

Sjors opened the pull request:
The `createwalletdescriptor` was introduced in #29130 to let users add a `tr()` descriptor to an existing SegWit wallet. The new `addhdkey` method from #29136 introduces a new potential workflow: start from a blank wallet, generate an HD and then add only the descriptors you need, e.g.:

```sh
bitcoin rpc createwallet TaprootMaxi blank=true
bitcoin rpc addhdkey
bitcoin rpc createwalletdescriptor bech32m
```

Before this PR the last line would fail, requiring the user to call `gethdkeys` and copy-paste the xpub.

This PR makes `createwalletdescriptor` a bit smarter so it just finds the `unused(KEY)` generated by `hdkey` and uses that.

If multiple `unused(KEY)` descriptors are present the user still has to pick one.

A potential followup is to make our [multisig tutorial](https://github.com/bitcoin/bitcoin/blob/master/doc/multisig-tutorial.md) slightly safer to use. Rather than creating a full wallet, the instruction could be changed to start with a blank wallet and only generate the default legacy descriptor. This avoids accidental use of single sig `p2sh-segwit` and `bech32` addresses.

(I might do that in #32784, which introduces some other improvements to the tutorial. The situation is still not great; ideally the `importdescriptors` RPC is enhanced to detect when an `xpub` is a child of an hd key and then treats it as if an xpriv was imported.)

--

The first two commits are quick followups for #29136.

Sjors committed:
wallet: add GetScriptlessSPKMs() helper

A helper method to obtain all unused(key) descriptor SPKMs.

Sjors committed:
rpc: make createwalletdescriptor smarter

When a wallet contains only an unused(KEY) descriptor, use it. Previously the user would have to call listdescriptors and manually specify it.

rkrux commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -883,11 +883,37 @@ static RPCHelpMan createwalletdescriptor()
+
+                    if (wallet_xpubs.empty()) {
+                        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "No HD key found. Please generate one with 'addhdkey' or import an active descriptor.");
+                    } else if (wallet_xpubs.size() > 1) {
+                        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Unable to determine which HD key to use. Please specify with 'hdkey'");
```
In e8b1440aa3ed7efe3a9c2d10afb05e7247fff1f6 "rpc: make createwalletdescriptor smarter"

Though this check here seems more thorough, but the presence of more than one `spkm` also seems sufficient to throw this error?

Sjors commented:
> Keeping it as an unused descriptor later might come across as confusing.

I agree, and it was also brought up here: https://github.com/bitcoin/bitcoin/pull/29136#issuecomment-1868202696 It's orthogonal to this PR.

adyshimony commented on test/functional/wallet_hd.py:
```diff
@@ -25,6 +27,56 @@ def set_test_params(self):
+        wallet = self.nodes[0].get_wallet_rpc("hdkey")
+
+        assert_equal(len(wallet.gethdkeys()), 1)
+
```
Cover the case where addhdkey is called with existing xprv. Expected behavior is to reject it as "HD key already exists"

```
existing_wallet_xprv = wallet.gethdkeys(private=True)[0]["xprv"]
assert_raises_rpc_error(-4, "HD key already exists", wallet.addhdkey, existing_wallet_xprv)
```

adyshimony commented:
Review and tested c8b6d4ee2abbd1271e44aa66da6079cee6323529 

Concept ACK, Approach ACK.

I think I have found a small issue where a xprv already in wallet via active descriptors, and addhdkey(xprv) is called from RPC with the same xprv.

Suggested a fix in the review comments.

Sjors committed:
test: check unused xprv descriptor pubkeys

Co-authored-by: adyshimony <6388409+adyshimony@users.noreply.github.com>

Sjors committed:
wallet: reject duplicate addhdkey xprvs

Co-authored-by: adyshimony <6388409+adyshimony@users.noreply.github.com>

Sjors commented on src/test/descriptor_tests.cpp:
```diff
@@ -1326,4 +1317,65 @@ BOOST_AUTO_TEST_CASE(descriptor_older_warnings)
+    // Check both only have one pubkey
+    std::set<CPubKey> prv_pubkeys;
+    std::set<CExtPubKey> prv_extpubs;
+    parse_pub->GetPubKeys(prv_pubkeys, prv_extpubs);
```
Thanks, fixed in a new commit (even though it pertains to #29136).

Sjors commented:
Rebased after #29136 landed and included suggestions from @adyshimony.

Sjors commented:
Rebased after https://github.com/bitcoin/bitcoin/pull/28333.

w0xlt commented:
Concept ACK

Sjors commented:
Rebased after #32784.

jeanpablojp commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -790,11 +790,37 @@ static RPCMethod createwalletdescriptor()
                     throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Unable to determine which HD key to use from active descriptors. Please specify with 'hdkey'");
+                } else {
+                    // Look for an unused(KEY) descriptor
+                    std::set<DescriptorScriptPubKeyMan*> spkms{pwallet->GetScriptlessSPKMs()};
```
`GetHDPubKeys(HDKeyFilter::UnusedKey)` already returns this set, and `derivehdkey` uses it further down in this same file. I swapped this block for a call to it, the wallet tests still pass and `GetScriptlessSPKMs()` goes away entirely. Is there a reason for the separate helper that I'm missing?

Sjors replied:
I dropped f336199336f619f6c361e56da06e36aa46f776c8, as it's indeed superseded by `GetHDPubKeys(HDKeyFilter::UnusedKey)` from #32784.

jeanpablojp reviewed (commented):
Concept ACK

I noticed that on a wallet with two active keys plus one from `addhdkey`, `derivehdkey` uses the unused one while `createwalletdescriptor` errors with "Unable to determine which HD key to use from active descriptors". Master does the same. But now that this RPC knows about unused(KEY), shouldn't that lookup also apply when there's more than one active key?

The `hdkey` hint is stale now too. It still says "The HD key used by all other active descriptors", where `derivehdkey`'s mentions both cases.

Sjors commented:
The rule and help text is now:

> The HD key used by all other active descriptors, or, if there are none, the HD key of an unused(KEY) descriptor.

In the ambiguous case @jeanpablojp describes, it's better to err on the side of failing, because descriptors can't be deleted. In that sense it's different from `derivehdkey`.

> The hdkey hint is stale now too.

Fixed

I also added a commit to use improve the multisig tutorial.
